### DISCUSSION TITLE
MagickImportImagePixels doesn't handle opacity import properly

### DISCUSSION THREAD

**Discussion by loops** [ImageMagick/ImageMagick#4187] [Help]
### Description
When importing raw pixel data from an X11 instance in the BGRO format,
the opacity byte is handled as an alpha (ie. 0 is transparent, not opaque).

### Steps to Reproduce
MagickImportImagePixels(wand,x,y,w,h,"BGRO",CharPixel,&pixels);

### System Configuration

- ImageMagick version:
git 497da390537afe2d6ed327aa6e3a508f566f4eb7

- Environment (Operating system, version and so on):
Fedora 32 Linux

Locally i've applied the attached patch as a temporary solution, but it only
handles my specific case; there are others I didn't patch because I'm not
sure this is actually the correct way to fix the issue.

[opacity_import.txt](https://github.com/ImageMagick/ImageMagick/files/5076264/opacity_import.txt)

**Comment by snibgo** [comment]
The alpha channel might represent _opacity_ or _transparency_. A zero value would mean "none of that quality". A value of 100% of Quantum would mean "entirely of that quality".

So Opacity=0 should mean "fully transparent".

**Comment by loops** [comment]
Hi snibgo,

I suspected that it was a mistake on my part, and am happy to learn how do this properly.  But i'm still confused.

The input data comes from the X11 server which has  0 = Opaque, 255 = Fully transparent.  My understanding was that you could import it with input format "BGRO", but it doesn't work.

So what you're saying is that this is actually an "A" alpha channel, because Alpha=0 should means fully opaque?  So I just tried importing the data as "BGRA", and yet when i then write the data out with MagickWriteImage(wand, "test.png") immediately without doing anything else, it appears as fully transparent in all image viewers and the Gimp.

Without my hacky patch,  "BGRA" and "BGRO" act the exact same way and both treat 0 = transparent.  As an aside, "BGRP" does properly ignore the alpha channel altogether.

**Comment by snibgo** [comment]
> The input data comes from the X11 server which has 0 = Opaque, 255 = Fully transparent.

So that channel should be called _transparency_.

Calling a channel "alpha" doesn't define what it does. Most file formats use alpha for _opacity_, so alpha=0 to mean "fully transparent" and alpha=maximum to mean "fully opaque". Many versions ago, ImageMagick defined alpha to mean transparency internally.

(But beware that some people just say the channel is "transparency", whichever way it operates.)

As you have seen, ImportImagePixels() is defined in pixel.c. Ignore the comments at the top as they are wrong; BGRA and BGRO do the same thing, and should do the same thing. There is no option for reading "BGRT", so you need to "-negate" the alpha channel, ie subtract it from QuantumRange.

In "BGRP", the "P" is for padding, ie the value has no meaning.

**Comment by loops** [comment]
snibgo,

That is most disappointing.   There seems to be no need to have two options "BGRA" and "BGRO" that do the same thing.  Why on earth are they handled as separate cases in the code, only to do the exact same thing?  It makes no sense.

I understand your argument from the language definition viewpoint, I agree that opacity means, how opaque something is, ie, 0 is transparent.  But from a functional viewpoint, it makes a lot more sense that these two cases are handled properly.  Just as they were handled in ImageMagick version 6.  This seems like reduced functionality for the sake of respecting a poor naming choice.

**Comment by snibgo** [comment]
>  Why on earth are they handled as separate cases in the code, only to do the exact same thing? It makes no sense.

I agree. But they used to do different things, many versions ago.

It might be sensible to add a "BGRT" option, that negates the alpha channel.

**Comment by loops** [comment]
Well the code literally has the two paths for GBRA and GBRO separated already, they're not handled as synonyms -- they just do the same thing.   So it would be easy enough to rename GBRO to GBRT and fix it to do the negation, and then make the GBRA path also handle GBRO as a synonym.

But this is a pain point for people porting code from Version 6 (forcing all GBRO to GBRT), and there is already a lot of resistance to moving to v7.  I would think this is an example of where we could make it easier rather than harder by simply adopting the old meaning of GBRO as it was.  But maybe that ship has sailed and it's already too late.

So if the only option is to add a BGRT, that would be a very helpful addition since otherwise it's an extra negate operation for every single import -- with the associated performance hit of having to traverse the image twice etc.  This is pretty common under Linux at least, but maybe not anywhere else.

**Comment by fmw42** [comment]
`convert image -channel a -negate +channel result`

will negate the alpha channel

If you have GBRO data and not RGBO, then you will need to swap channels. I know OpenCV use BGR, but I have not heard of any system using GBR.

**Comment by loops** [comment]
fmw42,  the data returned from the X11 server, (essentially a screenshot) is returned as BGRO.   It never hits the filesystem, it's all done in memory, never saved as an image file.  So I'm looking for a way to process those bytes in memory as it was done in V6, by simply specifying the import map of "BGRO".  Unfortunately it seems that the meaning has changed in V7, but i'm at a loss to understand why.
